File: MATLAB/Face Detection/Combined Algorithm of Shufflenet and MTCNN/face_detection_in_video_Shufflenet_MTCNN.m
% 1. Load the video
video = VideoReader('video_file_name.video_format');

% 2. Load pretrained shufflenet model
net = shufflenet;

% 3. Create a face detector object
detector = mtcnn.Detector();

% 4. Create a counter to keep track of detected faces
faceCount = 0;

% 5. Get the start time
tic;

% 6. Loop through each frame of the video
while hasFrame(video)
    % 7. Read the current frame
    frame = readFrame(video);

    % 8. Detect faces in the current frame
    [bboxes, scores, ~] = detector.detect(frame);
    % Increment the face counter
    faceCount = faceCount + numel(scores);

    % 9. Draw bounding boxes around the detected faces
    displayIm = insertObjectAnnotation(frame, 'rectangle', bboxes, scores, 'LineWidth', 2);
    imshow(displayIm);

    % 10. Loop through each detected face
    for i = 1:size(bboxes, 1)
        % Crop the detected face from the frame
        face = imcrop(frame, bboxes(i, :));
        % Increment the face counter
        faceCount = faceCount + 1;

        % 10. Save the detected face to a file
        filename = sprintf('%d.pgm', faceCount);
        imwrite(face, filename);
    end  % End of the loop for processing each detected face

end  % End of the loop for processing each frame

% 11. Get the end time
elapsedTime = toc;

% 12. Print the total number of detected faces
fprintf('Total Detected Faces: %d\n', faceCount);

% 13. Print the execution time
fprintf('Execution Time: %.3f seconds\n', elapsedTime);

File: Python/Face Detection/Combined Algorithm of Shufflenet and MTCNN/face_detection_in_video_Shufflenet_MTCNN.py
import cv2
from mtcnn import MTCNN
from google.colab.patches import cv2_imshow
import time

# Load the video
video = cv2.VideoCapture('video_file_name.video_format')

# Load pretrained shufflenet model (if needed)

# Create MTCNN face detector
detector = MTCNN()

# Create a counter to keep track of detected faces
face_count = 0

# Get the start time
start_time = time.time()

# Loop through each frame of the video
while video.isOpened():
    # Read the current frame
    ret, frame = video.read()

    if not ret:
        break

    # Detect faces in the current frame using MTCNN
    results = detector.detect_faces(frame)

    # Increment the face counter
    face_count += len(results)

    # Draw bounding boxes around the detected faces
    for result in results:
        bounding_box = result['box']
        score = result['confidence']
        cv2.rectangle(frame, (bounding_box[0], bounding_box[1]),
                      (bounding_box[0] + bounding_box[2], bounding_box[1] + bounding_box[3]),
                      (255, 0, 0), 2)
        cv2.putText(frame, f'Score: {score:.2f}', (bounding_box[0], bounding_box[1] - 10),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.5, (255, 0, 0), 2)

        # Crop the detected face from the frame
        face = frame[bounding_box[1]:bounding_box[1] + bounding_box[3],
                     bounding_box[0]:bounding_box[0] + bounding_box[2]]

        # Increment the face counter
        face_count += 1

        # Save the detected face to a file
        filename = f'{face_count}.pgm'
        cv2.imwrite(filename, face)

    # Display the frame
    cv2_imshow(frame)

    # Break the loop if 'q' is pressed
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

# Get the end time
elapsed_time = time.time() - start_time

# Release the video capture object
video.release()

# Print the total number of detected faces
print(f'Total Detected Faces: {face_count}')

# Print the execution time
print(f'Execution Time: {elapsed_time:.3f} seconds')
